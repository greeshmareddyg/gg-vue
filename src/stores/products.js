import { defineStore } from 'pinia'


export const productsStore = defineStore('products', {
  state: () => ({
    counter: 0,
    products: [],
    cart: [],
    searchText: '',
    filteredProducts: [],
    cartTotal: 0
  }),
  actions: {
    // Fetch Api data
    async fetchProductsFromAPI() {
      await fetch('https://dummyjson.com/products')
          .then(res => res.json())
          .then(json => {
            this.products = json.products;
          })
          .catch((error) => {
            console.log('error:', error)
            throw error;
          })

          console.log('api called')
          console.log('response', this.products)
    },

    addToCart(product) {
      this.cart.push(product)
      console.log('cart called ', this.cart)
    },

    removeFromCart(id) {
      console.log('product ID', id)
      this.cart = this.cart.filter((item) => item.id !== id)
    },
    setSearchText(input) {
      this.searchText = input
    },
  },
  mutations: {

  },
  getters: {
    // getters for different data rom products
    filteredProductsData: (state) => {
      return state.products.filter((product) => product.title.toLowerCase().includes(state.searchText.toLowerCase()))
    },
    getCartTotal: (state) => {
      let total = state.cart.reduce((sum, item) => sum + item.price, 0);
      return Math.round(total * 100) / 100
    },
    getCategoriesData: (state) => {
      console.log('get categories', state.products.reduce((acc,curr)=>(
        (acc[curr.category] = acc[curr.category] || []).push(curr), acc
      ), {}))
      let filteredcategories = state.products.reduce((acc,curr)=>(
        (acc[curr.category] = acc[curr.category] || []).push(curr), acc
      ), {})
      return filteredcategories
    }


  }

})