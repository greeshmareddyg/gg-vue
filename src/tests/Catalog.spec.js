import { mount } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { createTestingPinia } from "@pinia/testing";
import { describe, it, expect, vi, beforeEach } from "vitest";
import CatalogView from "@/views/Catalog.vue";
import { productsStore } from "@/stores/products";
import { setActivePinia, createPinia } from "pinia";

const routes = [{ path: "/product/:id", name: "ProductView" }];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

describe("Unit test case for CatalogView", () => {
  let store;

  beforeEach(() => {
    const pinia = createPinia();
    setActivePinia(pinia);
    store = productsStore();
    store.$state.products = [
      { id: 1, title: "Product 1" },
      { id: 2, title: "Product 2" },
      { id: 3, title: "Product 3" },
    ];
    store.$state.searchText = "";
    store.$state.filteredProductsData = store.$state.products;
  });

  it("filters products based on search text", async () => {
    const setSearch = vi.fn();
    const wrapper = mount(CatalogView, {
      global: {
        plugins: [router, setSearch, createTestingPinia({ createSpy: vi.fn })],
      },
    });

    const input = wrapper.find("v-text-field");
    expect(input.exists()).toBe(true);

    await input.trigger("update:modelValue", "Product 1");
    await wrapper.vm.$nextTick();
  });
});
