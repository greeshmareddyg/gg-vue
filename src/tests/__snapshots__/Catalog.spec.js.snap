// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Catalog.vue > renders the catalog view 1`] = `
<div
  class="products-list grid-container"
  data-v-e5ae2966=""
>
  <div
    class="column"
    data-v-e5ae2966=""
  >
    <v-text-field
      class="search-input"
      clearable=""
      data-v-e5ae2966=""
      label="Search"
      modelvalue=""
    />
    <!--v-if-->
    <!--v-if-->
    <span
      data-v-e5ae2966=""
    >
      0 results 
    </span>
    <v-row
      data-v-e5ae2966=""
      no-gutters=""
    >
      
      
    </v-row>
  </div>
</div>
`;
